### ISSUE TITLE
SPFx 1.12 - TypeError: Object(...) is not a function

### ISSUE THREAD

**Issue by coreyroth** [pnp/pnpjs#1654]
### Category
- [ ] Enhancement
- [X] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [  2.3.0 (started with 2.2.0)       ]

Please specify what version(s) of SharePoint you are targeting: [   SharePoint Online with SPFx 1.12      ]

*If you are not using the latest release, please update and see if the issue is resolved before submitting an issue.*

### Expected / Desired Behavior / Question
After upgrading to SPFx 1.12 this morning, I have found that the following error occurs now when I use methods like sp.web.lists.GetById() or sp.hubSites.getById().  I have tried this with version 2.2.0 and upgraded to 2.3.0.  

### Observed Behavior
I get the following error when using SPFx 1.12.

TypeError: Object(...) is not a function
    at factory (invokable-binder.js:17)
    at applyFactoryExtensions (invokable-extensions.js:86)
    at invokable-binder.js:26
    at index.js:12
    at SPRest.9l4K.SPRest.childConfigHook (rest.js:67)
    at SPRest.get [as hubSites] (index.js:10)
    at ApplicationInsightsService.<anonymous> (ApplicationInsightsService.ts:999)
    at Generator.next (<anonymous>)
    at application-customizer_8bf88ad57c5ebffbf016.js:18455
    at new Promise (<anonymous>)

When I revert to my SPFx 1.11 builds, this error does not occur.

### Steps to Reproduce
Here's a short code snippet:

`   try {
      const hubSiteInfo: IHubSiteInfo = await sp.hubSites.getById(hubSiteId)();
      return hubSiteInfo;
    }
    catch (error) {
     throw error;
    }`

I don't think I am doing anything wrong.  I have checked all of my dependencies.  I am going to upgrade another project and see if the same issue occurs with my PnPJS calls there as well.

**Comment by koltyakov** [comment]
Hey @coreyroth! What's the Browser? IE? Then polyfills is an answer.

**Comment by coreyroth** [comment]
Edge 90 on Mac OS.  I've already got the polyfills included for backwards compatibility.

**Comment by koltyakov** [comment]
Oh, that's interesting. NO polyfills needed for new Edge.
The selective imports then to check - https://pnp.github.io/pnpjs/concepts/selective-imports/.
Could you check is the [preset](https://pnp.github.io/pnpjs/concepts/selective-imports/#sp) works?

**Comment by coreyroth** [comment]
Ok.  So I replaced the following selective imports:

`import { sp } from "@pnp/sp";
import "@pnp/sp/webs";
import "@pnp/sp/lists";
import "@pnp/sp/items";
import "@pnp/sp/comments";
import "@pnp/sp/hubsites";`

with 

`import { sp } from "@pnp/sp/presets/all";`

and it yields the same results.

**Comment by patrick-rodgers** [comment]
Can you share a full repro? Want to make sure we are able to see what you're seeing. Did some initial testing with 1.12 and things seemed ok, but didn't go very deep.

Out of curiosity, if you remove our polyfills does it work? Wondering if there is a conflict there as I didn't have polyfills when I did my admittedly quick tests.

**Comment by coreyroth** [comment]
I'll try removing the polyfills real quick.  I'm also upgrading my second project to see if it also occurs there.  Definitely could be user error on my part somewhere along the ways when I upgraded.  Upgraded using m365 cli.

**Comment by coreyroth** [comment]
Ok.  Project 2 works fine when it makes a call to get information about a Site.  It also has polyfills included.  Let me try changing TypeScript versions on the first project.

**Comment by coreyroth** [comment]
Back to project 1.  I downgraded to TypeScript 3.7.  I tried removing polyfills.  I removed node_modules / package-lock.json and reinstalled dependencies.  Another difference I thought of was I am using "es2015" in the lib section of my tsconfig.json in this project.  Maybe I should start a new branch and try the upgrade again to see if missed something along the way in the upgrade.

**Comment by coreyroth** [comment]
I can reproduce it in a new project using an Application Customizer.  Simple call to get IHubSiteInfo in this case.  Here is the link.

[Repo](https://github.com/coreyroth/SPfx1.12-PnPJSError)

**Comment by DavidDavidDev** [comment]
I am experiencing the same error message, Mac OS, Chrome, SPFx 1.12, gulp 4.0.2, node 12.21.0

pnp/sp packages 2.3

Similarly, I upgraded my environment (yo, gulp, @microsoft/generator-sharepoint), started a new project. Initial Hello World worked fine. 

Fails during the first PnP list service call. Tried different import methods as described above but it didn't resolve the issue.

**Comment by RicoNL** [comment]
I am experiencing the same issue. Same scenario as DavidDavidDev described, except Windows 10 instead of MacOS. 

This was working fine with SPFx 1.11

**Comment by coreyroth** [comment]
I'm happy that others were able to reproduce this issue.

**Comment by saikatojha** [comment]
I am also experiencing the same issue with SPFx 1.12.0 version. Tried with Node JS 10.24.0 and 12.21.0

**Comment by Ofer-Gal** [comment]
Same here when running something simple as:
```
        let web = Web(window.location.origin + "/teams/" + projectNumber);
```
StackOverfolw had a mention of it to do with hooks so I tried with 16.8.5 but that did not fix it.

**Comment by FredrikEkstroem** [comment]
Some observations, not sure if relevant since I still haven't got my head around the whole ecosystem :-)

it but it looks like the issue is related to the __spreadArray function in tslib that was added in the last version (2.1).
The source code https://github.com/pnp/pnpjs/blob/version-2/packages/odata/queryable.ts  uses
return JSON.stringify([...(<Map<string, string>>value)]);

That is translated to return JSON.stringify(__spreadArray([], __read(value))); in the queryable.js in the npm package. However that function doesn't exist in the version of tslib that is used by default in spfx. 

Manually updating the tslib package to 2.1 breaks the whole webpart project.

spfx 1.12 runs "TypeScript to v3.7 (via @microsoft/rush-stack-compiler-3.7 v0.2.3)". 
recommended installations for tslib are:
# TypeScript 3.9.2 or later
npm install tslib
# TypeScript 3.8.4 or earlier
npm install tslib@^1

**Comment by patrick-rodgers** [comment]
Thanks @FredrikEkstroem - I am coming to the same conclusion. Appears to be something with tslib version mismatches. We aren't getting "our" version 2.1.0 but instead it appears we are getting some of 2.1.0 and some of 1.10.0 (SPFx version). I haven't been able to figure out why exactly that is happening yet. It does appear that the error isn't our code exactly but rather some interaction with the bundling and tslib. Specifically this transpiled code:

```JavaScript
var r = Object.assign(function () {
    var ags = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        ags[_i] = arguments[_i];
    }
    return invoker.call.apply(invoker, __spreadArray([r], __read(ags)));
}, new (constructor.bind.apply(constructor, __spreadArray([void 0], __read(as))))());
```

Error occurs in this line: `new (constructor.bind.apply(constructor, __spreadArray([void 0], __read(as))))());`

- as => [""]
- __read => import from tslib, appears to be working (I can step into it)
- __spreadArray => doesn't exist and is the source of the error
- constructor => defined in the way it always has been (i.e. no changes to code affecting this from 2.2.0 to 2.3.0)

**Comment by Ofer-Gal** [comment]
Tried to test the options :-) did:
```
npm i @microsoft/rush-stack-compiler-3.9
"extends": "./node_modules/@microsoft/rush-stack-compiler-3.9/includes/tsconfig-web.json", in tsconfig.json
npm install tslib
```

then tried in the work bench. but got the error:
```
[SPLoaderError.loadComponentError]:
Failed to load component "e509db72-e72d-447b-ae6c-8cbc515d71fd" (AceFileMangerWebPart). Original error: Manifest not found for component id "01c4df03-e775-48cb-aa14-171ee5199a15" and version "2.1.0".

INNERERROR:
Manifest not found for component id "01c4df03-e775-48cb-aa14-171ee5199a15" and version "2.1.0".
CALLSTACK:
Error
    at t [as constructor] (https://modern.akamai.odsp.cdn.office.net/files/sp-client/sp-webpart-workbench-assembly_en-us_ab63dd233c63fab8a494610301b143a3.js:65:19890)
    at new t (https://modern.akamai.odsp.cdn.office.net/files/sp-client/sp-webpart-workbench-assembly_en-us_ab63dd233c63fab8a494610301b143a3.js:65:169952)
    at Function.e.buildErrorWithVerboseLog (https://modern.akamai.odsp.cdn.office.net/files/sp-client/sp-webpart-workbench-assembly_en-us_ab63dd233c63fab8a494610301b143a3.js:65:207271)
    at Function.e.buildLoadComponentError (https://modern.akamai.odsp.cdn.office.net/files/sp-client/sp-webpart-workbench-assembly_en-us_ab63dd233c63fab8a494610301b143a3.js:65:203316)
    at https://modern.akamai.odsp.cdn.office.net/files/sp-client/chunk.systemjs-component-loader_en-us_ec51c37c549687dc4960.js:1:65979
```

I hope it helps you guys to resolve the problem.
I have no clue :-)  
Thank you anyway.

**Comment by patrick-rodgers** [comment]
Thanks @Ofer-Gal, I hit that too. Appreciate the info.

**Comment by patrick-rodgers** [comment]
For folks having trouble please install the latest beta version and let us know if the issue is resolved. As a stop gap I am not importing the helpers so they are inline. My testing now works without issue. If this works for others we can release a 2.3.1 version with this fix in place while we work on the strangeness around the bundling/transpile/etc happening with SPFx. Feels like there is a lot of trial and error there so want to get an interim fix out. Thanks!

**Comment by coreyroth** [comment]
Ok, I tried 2.3.1@beta0 and I am now seeing a different error on my tenants.

_ Error making HttpClient request in queryable [404]  ::> 
    at new HttpRequestError (parsers.js:178)
    at Function.<anonymous> (parsers.js:193)
    at step (parsers.js:45)
    at Object.next (parsers.js:26)
    at fulfilled (parsers.js:17)_

**Comment by PathToSharePoint** [comment]
For the record the manifest issue is not specific to pnpjs. [I reported it via twitter ](https://twitter.com/Path2SharePoint/status/1372977896713691137)in a non-pnpjs scenario.

**Comment by RicoNL** [comment]
I tried with 2.3.1-beta0 and seems to work now for me:

let items: IItems = await sp.web.lists.getByTitle("Demo").items();

**Comment by coreyroth** [comment]
With 2.3.1-beta0, I am seeing 404 errors on API calls to list items and to hub sites.  Look at this example of the URL.  It has "sites" in front of "/_api".

`https://mytenant.sharepoint.com/sites/_api/hubsites/GetById?hubSiteId=%274e4fbb7a-d91c-49a4-9d17-2248cec496e7%27`

**Comment by patrick-rodgers** [comment]
@coreyroth - please open a new issue for the new error you are reporting so we can more easily track what's being reported. Thanks!

**Comment by chr-sad** [comment]
I could also reproduce the same error on Windows 10 + Chrome with SPFx 1.12.0 which was working on SPFx 1.11.0. The suggested **version 2.3.1-beta0 seems to work for me as well.** I have not tested our whole solution in detail, but so far no problems found.

**Comment by FredrikEkstroem** [comment]
@patrick-rodgers  I can also confirm 2.3.1-beta0 works for all scenarios I have tested.

OT: @Ofer-Gal  I am also getting the "Manifest not found for component id" a lot, not just connected to pnp js, and that's also connected to tslib. I haven't figured out simple reproduction steps yet though. https://github.com/pnp/pnpframework/issues is the right place for reporting that?

**Comment by PathToSharePoint** [comment]
@FredrikEkstroem The recommendation I got from @VesaJuvonen was to post on https://github.com/SharePoint/sp-dev-docs/issues

Reference: https://twitter.com/vesajuvonen/status/1373577775098032128

**Comment by Ofer-Gal** [comment]
@FredrikEkstroem I reported it but wa asked for steps to reproduce. but I can't because it doesn't happen to me but to me users. I only know about it because of logging.

**Comment by patrick-rodgers** [comment]
Thanks all for the feedback on the beta. After I get a chance to talk this morning with @juliemturner we will likely push out a 2.3.1 patch release with all our tslib refs inlined. That is a _temporary_ solution to unblock folks using this library. It will ONLY address the OP issue in this thread `TypeError: Object(...) is not a function`.

For general issues you are encountering with SPFx 1.12 please report them to the [dev-docs issue list](https://github.com/SharePoint/sp-dev-docs/issues). This include other issues with tslib not associated with PnPjs.

Appreciate everyone's help ensuring we are capturing any problems you see so they can be properly tracked and resolved. Thanks!!

**Comment by PathToSharePoint** [comment]
May I suggest that people opening the tslib issue drop a note here? To prevent MS folks being flooded with duplicates. I'll work on a sample myself but only after I switch back to Nodejs 12.

**Comment by FredrikEkstroem** [comment]
tslib issue added, see link above. If you have anything else to add or just want to confirm you have the same issue, use that one.

**Comment by FredrikEkstroem** [comment]
According to @VesaJuvonen SPFX 1.12.0 is to be "considered as deprecated", and a fix will be coming later, [see tweet](https://twitter.com/vesajuvonen/status/1374620639080894464).

Is this issue part of that fix?

**Comment by patrick-rodgers** [comment]
Based on what is happening with v1.12 of SPFx we are NOT going to publish our interim fix as a version of PnPjs. It remains as a beta version for anyone that may happen to need it, but we will wait and see how the SPFx 1.12.? updated package works and addresses this issue. My understanding is that tslib issues affected many libraries and should be fixed in the upcoming release.

Once that is out we can retest this library to ensure integration works and should be sorted.

Thanks again everyone for the comments - closing this issue as moot since the SPFx version is deprecated.
